modules/framework/galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/routes/AuthRoute.java

```diff
@@ -78,35 +81,35 @@ public HttpServletResponse handleGetRequest(String pathInfo, QueryParameters que
             HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException, FrameworkException {
 
         logger.info("AuthRoute: handleGetRequest() entered.");
-        HttpSession session = request.getSession(true);
         try {
-            String clientId = queryParams.getSingleString("client_id", null);
-            String clientCallbackUrl = queryParams.getSingleString("callback_url", null);
+            String clientId = sanitizeString(queryParams.getSingleString("client_id", null));
+            String clientCallbackUrl = sanitizeString(queryParams.getSingleString("callback_url", null));
 
             // Make sure the required query parameters exist
             if (clientId == null || clientCallbackUrl == null || !isUrlValid(clientCallbackUrl)) {
                 ServletError error = new ServletError(GAL5400_BAD_REQUEST, request.getServletPath());
                 throw new InternalServletException(error, HttpServletResponse.SC_BAD_REQUEST);
             }
 
-            // Store the callback URL in the session to redirect to at the end of the authentication process
-            session.setAttribute("callbackUrl", clientCallbackUrl);
+            // Create a random ID to identify this auth request
+            String stateId = UUID.randomUUID().toString();
 
             // Get the redirect URL to the upstream connector and add it to the response's "Location" header
-            String authUrl = oidcProvider.getConnectorRedirectUrl(clientId, AuthCallbackRoute.getExternalAuthCallbackUrl(), session);
+            String authUrl = oidcProvider.getConnectorRedirectUrl(clientId, AuthCallbackRoute.getExternalAuthCallbackUrl(), stateId);
             if (authUrl != null) {
                 logger.info("Redirect URL to upstream connector received: " + authUrl);
                 response.addHeader(HttpHeaders.LOCATION, authUrl);
 
+                // Store the callback URL in the DSS to redirect to at the end of the authentication process
+                dssService.put(stateId + DSS_CALLBACK_URL_PROPERTY_SUFFIX, clientCallbackUrl, AUTH_DSS_STATE_EXPIRY_SECONDS);
```

Comment:
Could a malicious user flood the DSS with auth-related state by kicking off a lot of auth requests and not completing them? Rate-limiting by IP (e.g. via nginx Ingress) would help mitigate this. Is there anything else that could be done?

Reply:
I think rate limiting is all we can do really. And sort out something which de-frags etcd every week.

Reply:
I think I'd be more worried about 2 parallel pods generating the same random ID, so one person gets the others' JWT back.

Reply:
Perhaps we should add the IP address of the requestor to the key ?

Reply:
That makes sense - added the IP address of the client to the random ID key in https://github.com/galasa-dev/galasa/pull/149/commits/0cef09e4b9e921821c6d8e54ca16988d6569b448